MSVC Compatibility

jseyster opened the pull request.

mdashti committed:
Removed all Non-SpiderMonkey Files

mdashti committed:
Removed Rust dependencies from the build scripts

mdashti committed:
Removed Wasm testing support + Updated the tests that depend on Wasm

mdashti committed:
Replaced `encoding_rs` with its counterpart C++ implementation

mdashti committed:
Fixed unused variable warnings in the production build.

mdashti committed:
Move a static variable inside function impl

mdashti committed:
Added a missing header + Fixed an erroneous noexcept

mdashti committed:
Used ICU by default, as it is needed for the implementation of `encoding_rs` replacement

mdashti committed:
Fixed the C++ implementation bugs in the `encoding_rs` replacement code

mdashti committed:
Fixed some corner case bugs in `mfbt` re Latin1 and Utf8 + Nitz

mdashti committed:
Added unit-tests for `mfbt` changes re Latin1, Utf8 and TextUtils

mdashti committed:
Making the code-base C++20 compatible

mdashti committed:
Nitz.

mdashti committed:
Added `AUTHORS` and `README.txt` back as suggested by @paroski

mdashti committed:
Nitz

mdashti committed:
Nitz

mdashti committed:
Removed Rust from SpiderMonkey Repo (#4)

Removed Rust dependencies from the build scripts

mdashti committed:
Merge branch 'spidermonkey-esr91.3-cpp-only-merged-prs' into spidermonkey-esr91.3-cpp-only-add-unit-tests

# Conflicts:
#	config/rules.mk
#	js/moz.configure
#	js/src/js-standalone.mozbuild
#	js/src/moz.build
#	js/src/tests/shell/mailbox.js
#	js/src/wasm/WasmJS.h

mdashti committed:
Nitz.

mdashti committed:
Nitz suggested by @paroski

mdashti committed:
Reverted the `noexcept` marker removal pointed out by @paroski

mdashti committed:
Merge branch 'spidermonkey-esr91.3-cpp-only-merged-prs' into spidermonkey-esr91.3-cpp-only-add-unit-tests

mdashti committed:
Revert the change in `.gitignore`

mdashti committed:
Fixed the compilation issues after the merge.

mdashti committed:
Merge remote-tracking branch 'origin/spidermonkey-esr91.3-cpp-only-add-unit-tests' into spidermonkey-esr91.3-cpp20-compatibility

mdashti committed:
Nitz.

mdashti committed:
Lint changes

mdashti committed:
Merge branch 'spidermonkey-esr91.3-cpp-only-add-unit-tests' into spidermonkey-esr91.3-cpp20-compatibility

mdashti committed:
Applied PR comment

mdashti committed:
Applied PR comment by @BillyDonahue

mdashti committed:
Applied PR comment by @paroski

mdashti committed:
Applied PR comments by @paroski

mdashti committed:
Merge branch 'spidermonkey-esr91.3-cpp-only-add-unit-tests' into spidermonkey-esr91.3-cpp20-compatibility

mdashti committed:
Moved one more function and some headers to `Utf8.cpp`

mdashti committed:
Merge branch 'spidermonkey-esr91.3-cpp-only-add-unit-tests' into spidermonkey-esr91.3-cpp20-compatibility

jseyster committed:
Add MSVC alternative to __atomic_compare_exchange_n

jseyster committed:
Fix use of non-standard (in C++17) designated initializer

This change is required for our Windows build: "designated
initializer" syntax is a non-standard extension supported by Clang and
g++ but not by MSVC.

Note that this syntax _is_ standard in C++20, and we can revert this
change once we switch to C++20 for all builds.

jseyster committed:
MainThreadOrGCTaskData implicit unwrap workaround for MSVC

On Clang and g++, a MainThreadOrGCTaskData<gc::MarkColor> gets
implicitly converted to a gc::MarkColor, but this conversion sometimes
needs to be explicit for MSVC.

jseyster committed:
Replace result_of_t with invoke_result_t

When compiling with MSVC, result_of_t causes a deprecation error.

jseyster committed:
Simplify calls to mozjs StringBuffer::append(const char (&array)[ArrayLength])

In Clang/g++, type deduction figures out that both sides of the
ternary operator (see the diff) can be a char[8] array, which allows
overload resolution to choose the
StringBuffer::append(const char (&array)[ArrayLength]) overload, where
ArrayLength is a template parameter with value 8.

MSVC's type deduction cannot make this inference, however, so we need
to interchange the outer StringBuffer::append() call with the inner
ternary operator, simplifying inference. The resulting code is
arguably more robust: before this change, the code was only legal
because the ".getter" and ".setter" strings happen to have the same
length.

jseyster committed:
Do not use MSVC std::has_unique_object_representations

As noted in the comments, MSVC's
std::has_unique_object_representations rejects some data types that it
permits in the Clang/g++ implementations: namely,
js::ImmutableScriptData and js::frontend::ParserAtom.

It appears that MSVC does not consider these classes to be
TriviallyCopyable, which is one of the requirements for
std::has_unique_object_representations:
https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable

jseyster committed:
Explicitly capture 'ColumnChunkLength' value

Clang/g++ allow implicit capture for this variable, presumably because
it is defined as a constexpr, but MSVC does not.

jseyster committed:
Do not mark ConcreteTraceable::trace() as deleted in MSVC

jseyster committed:
Fall back to switch-based interpreters in MSVC

Mozjs has two bytecode interpreters, each of which operates in
"threaded" mode* by default: one for the Javascript VM and one for the
regular expression state machine (imported from V8).

In a "threaded" interpreter, each instruction has a label; an array
stores the address of each label, indexed by the instruction opcode;
and each instruction finishes its execution by looking up the label
for the next instruction and jumping to that label with a "computed
goto."

This style of interpreter requires special && syntax (no relation to
r-value references) to get the address for each label, as well as
support for using a runtime value as the target for a goto
statement. These are supported as extensions in Clang and gcc, but a
fallback switch-based approach is necessary for MSVC, which does not
support these extensions.

Preprocessor magic makes it possible to write a single interpreter
that can use either approach, depending on a small set of macro
definitions. The V8 regex interpreter already supports both
approaches, and this patch chooses the fallback when compiling on
Windows. The mozjs Javascript interpreter supported a switch-based
fallback in a _previous_ release, but that fallback was removed,
because none of mozjs's officially supported compilers need it. This
path reverses that removal, porting the switch-based interpreter
forward to this release.

*Note: "threaded" here does not refer to multi-threaded execution. The
interpreter still runs on a single thread.

jseyster committed:
Disambiguate 'ToStringSlow' function

It is not possible to get the address of an overloaded function by its
name, because the compiler would not know which overload to
choose. This patch adds an alias for the 'ToStringSlow' overload that
needs to be added to the VMFUNCTION_LIST.

It seems that Clang and g++ do not consider the other overloads of
'ToStringSlow' (in Conversions.h), because they are template
instantiations. MSVC does consider these overloads, however, resulting
in an error when compiling the VMFUNCION_LIST without using the
disambiguating alias.

jseyster committed:
Eliminate identity reinterpret_cast

On Clang/g++, reinterpret_cast from uintptr_t->uintptr_t is a no-op,
but on MSVC, it produces a compiler error.

jseyster committed:
Explicitly use ASCII version of 'GetModuleHandle' function

The 'GetModuleHandle' macro chooses between 'GetModuleHandleW' ("wide"
unicode string input) and 'GetModuleHandleA' (ASCII string input)
depending on the environment. In this case, however, the input is an
ASCII string literal, so the it is better to explicitly choose the "A"
variant, regardless of the environment.

jseyster committed:
Disambiguate reference to js::detail::TypeIsGCThing

jseyster committed:
Use _ReturnAddress() intrinsic for stack walking in MSVC

jseyster committed:
Work around MSVC ICE

Without these reinterpret_casts to simplify template deduction,
parsing of this call to AutoCallVM::call() triggers an internal
compiler error in MSVC 19.30.30706 (tested).

Weirdly, only the NotEqual case triggers the ICE. Both EqualityKind
cases are modified here, however, for symmetry.

The workaround was discovered using a highly sophisticated "take wild
guesses until something works" approach. Unfortunately, the resulting
code is not valid in Clang and g++, so conditional compilation is
necessary to support all compilers.

jseyster committed:
Files that include debugger/Debugger.h should also include other debugger headers

The Debugger class in Debugger.h has several DebuggerWeakMap
members. In Clang/g++, the compiler only instantiates the methods of
the DebuggerWeakMap's type parameters if the compile unit includes a
call to a function that mutates the DebuggerWeakMap object, which
makes some of these header files unnecessary.

MSVC, however, _always_ instantiates those members, and the
instantiations fail without the class definitions in Environment.h,
Frame.h, Script.h and Source.h.

jseyster committed:
Move PropertyFlags::defaultDataPropFlags definition out of header file so that it does not cause duplicate definition errors in MSVC

BillyDonahue commented on js/src/frontend/StencilXdr.h:
```diff
@@ -27,7 +27,11 @@ namespace frontend {
 struct CanCopyDataToDisk {
   // Check that the object is fully packed, to save disk space.
-#ifdef __cpp_lib_has_unique_object_representations
+  // Note that we do not use the MSVC std::has_unique_object_representations,
```
[optional] If there's an MSVC bug to track, that would be useful to cite here so we know when to remove this workaround.

jseyster replied:
Agreed. I haven't filed any bugs yet, but I'd like to when I have a spare day to set up some minimal examples. Have you submitted MSVC bugs before? I'd be interested in the process. Thanks!

BillyDonahue replied:
Yes I've done it. You do it here. https://developercommunity.visualstudio.com/home

jseyster replied:
Thanks for the info! I will try to get feedback submitted after these changes are ready.

BillyDonahue commented on js/src/vm/Interpreter.cpp:
```diff
@@ -1882,31 +1882,53 @@ void js::ReportInNotObjectError(JSContext* cx, HandleValue lref,
-#undef TRAILING_LABEL
+# undef TRAILING_LABEL
   };
+#else
```
It would be nice to label the else/endif so we don't get lost.

    #else  // have computed goto
    ...
    #endif // have computed goto

jseyster replied:
Good idea, thanks.

BillyDonahue commented on js/src/vm/Interpreter.cpp:
```diff
@@ -1882,31 +1882,53 @@ void js::ReportInNotObjectError(JSContext* cx, HandleValue lref,
+# define INTERPRETER_LOOP()
+# define CASE_PSEUDO(OP) label_##OP:
+# define CASE(OP) label_##OP:
+# define DEFAULT() \
```
Many of these lines didn't need to change. Let's not insert a space after `#` in these directives to avoid noise.
If that's what they had upstream we can conform to it.

jseyster replied:
I messed up here. I thought I had run this file through Firefox's `.clang-format`, but it turns out I didn't. After formatting, these lines actually have two spaces following the `#`. Personally, I think the extra readability from indenting nested directives and keeping formatting consistent is worth the diff noise, but I could remove the indentations if you think we should aggressively limit the diff size.

For reference, I copied this code from the older version of mozjs in our current master branch:
https://github.com/mongodb/mongo/blob/581c58c475a872e25b2e3bf7cf5ccd52425ef7c7/src/third_party/mozjs/extract/js/src/vm/Interpreter.cpp#L1755-L1789

BillyDonahue commented on js/src/vm/Interpreter.cpp:
```diff
@@ -1882,31 +1882,53 @@ void js::ReportInNotObjectError(JSContext* cx, HandleValue lref,
+ * and should work on any standard-compliant C++ compiler.
+ */
+# define INTERPRETER_LOOP()       the_switch: switch (switchOp)
+# define CASE_PSEUDO(OP)          case OP:
```
`PSEUDO` is unclear and specific to the only `OP` it happens to be used with. Maybe `CASE_NOT_COMPUTED` ?

jseyster replied:
Done. Thanks for the suggestion.

BillyDonahue commented on js/src/vm/Interpreter.cpp:
```diff
@@ -1882,31 +1882,53 @@ void js::ReportInNotObjectError(JSContext* cx, HandleValue lref,
+    JS_END_MACRO
+
+    // This variable is effectively a parameter to the_switch.
+    jsbytecode switchOp;
```
This should not be indented.

jseyster replied:
Formatting this file with the mozjs `.clang-format` corrected this indentation for us. Nice catch, though!

BillyDonahue commented on js/src/wasm/WasmCode.h:
```diff
@@ -663,10 +667,18 @@ class JumpTables {
+
+#ifdef _MSC_VER
+    int64_t expected = 0;
+    (void)_InterlockedCompareExchange64_np(reinterpret_cast<intptr_t*>(&jit_.get()[i]),
```
I can't find documentation for the `_np` version of this intrinsic?

https://docs.microsoft.com/en-us/previous-versions/ttk2z1ws(v=vs.85)

Maybe we should call through the public `InterlockedCompareExchange64` instead, and let the compiler macros find the right C API "overload" for us?

jseyster replied:
Thanks for this suggestion!

Good news:
I agree that this API function is more robust than relying on an intrinsic, and your suggestion to use `InterlockedCompareExchange64` led me to `InterlockedCompareExchangePointer`, which I think is even better for this purpose.
https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedcompareexchangepointer

Base news:
Both API functions require including `winnt.h`, which adds a bunch of `$define`s that conflict with identifiers throughout mozjs. I think the best way to deal with that would be to sequester `JumpTables::setJitEntryIfNull` to its own file. I've tried that out here. Let me know what you think.

BillyDonahue commented on js/src/wasm/WasmBaselineCompile.cpp:
```diff
@@ -2264,10 +2264,12 @@ class BaseStackFrame final : public BaseStackFrameAllocator {
-    } bits = {.f32 = imm};
+
+      StoredFloat32(float f32) : f32(f32) {}
+    } bits{imm};
```
remove blank.

Anonymous unions have certain special properties that could be erased by naming it. I think we can avoid doing so. A brace init of an anonymous union initializes its first member. So we can just swap the fields here:

```suggestion
    union {
      float f32;
      int32_t i32;
    } bits = {imm};
```

jseyster replied:
Very interesting! Thanks for catching this one and for the helpful workaround.

BillyDonahue commented on js/src/wasm/WasmBaselineCompile.cpp:
```diff
@@ -2280,10 +2282,12 @@ class BaseStackFrame final : public BaseStackFrameAllocator {
 
   void storeImmediateF64ToStack(double imm, uint32_t destHeight,
                                 Register temp) {
-    union {
+    union StoredFloat64 {
       int64_t i64;
       double f64;
-    } bits = {.f64 = imm};
+
+      StoredFloat64(double f64) : f64(f64) {}
+    } bits{imm};
```
We can just swap the fields here.
```c++
    union {
      double f64;
      int64_t i64;
    } bits = {imm};
```

jseyster replied:
Done.

BillyDonahue commented on js/src/wasm/WasmFrame.h:
```diff
@@ -225,6 +225,8 @@ class DebugFrame {
       uint32_t hasSpilledRefRegisterResult : MaxRegisterResults;
     };
     uint32_t allFlags;
+
+    constexpr Flags(uint32_t allFlags) : allFlags(allFlags) {}
```
If we just put `allFlags` first in the union, we can initialize `_flags{expr}` with a `uint32_t expr`.

jseyster replied:
Done.

BillyDonahue commented on js/src/wasm/WasmFrame.h:
```diff
@@ -317,7 +319,7 @@ class DebugFrame {
   static constexpr uint32_t hasSpilledRegisterRefResultBitMask(size_t n) {
     MOZ_ASSERT(n < MaxRegisterResults);
-    union Flags flags = {.allFlags = 0};
+    union Flags flags{0};
```
Alternatively I believe this could just change `{0}` to `{}` and rely on value-initialization of the struct?

jseyster replied:
Also a good idea. Between the two alternatives, I decided to go with reordering the Flags union so that we can initialize using a 0-value uint32_t. My reasoning is that it's conceivable that the compiler could init _just_ the bits that are specified in the struct to 0 and leave random junk in the other bits. Hard to imagine that happening in practice, but it's the best logic I could think of for choosing between the possible ways to initialize this variable.

BillyDonahue commented on js/src/frontend/TokenStream.cpp:
```diff
@@ -834,7 +834,7 @@ uint32_t TokenStreamAnyChars::computePartialColumn(
                                     this,
 #endif
-                                    start, limit,
+                                    start, limit, ColumnChunkLength,
```
[nit] Maybe put it on its own line so we diff more cleanly.

Maybe add a `static` to the definition of `ColumnChunkLength` instead? It shouldn't be necessary to add it to the capture list of this lambda.

jseyster replied:
Adding the `static` worked. Thanks!

BillyDonahue commented on js/src/jit/VMFunctionList-inl.h:
```diff
@@ -262,7 +262,7 @@ namespace jit {
   _(ThrowUninitializedThis, js::ThrowUninitializedThis)                        \
   _(ToBigInt, js::ToBigInt)                                                    \
-  _(ToStringSlow, js::ToStringSlow<CanGC>)
+  _(ToStringSlow, js::ToStringSlowForVM<CanGC>)
```
Instead of modifying the string API for this special case, we can, I think, define ToStringSlowForVM here in this file instead?

jseyster replied:
Done.

BillyDonahue commented on js/src/vm/StringType.h:
```diff
@@ -1788,6 +1788,18 @@ template <AllowGC allowGC>
     JSContext* cx, typename MaybeRooted<Value, allowGC>::HandleType arg);
 
+/*
+ * Same as the 'ToStringSlow' function. The VMFUNCTION_LIST in
+ * VMFuncionList-inl.h cannot include any overloaded functions, so this name is
+ * provided for use in that list. ('ToStringSlow' has an overload in
+ * Conversions.h.)
```
That's not a concern local to this file. I think it belongs in the VMFunctionList-inl.h to which it's pertinent?

jseyster replied:
Agreed.

BillyDonahue commented on js/src/jit/CacheIRCompiler.cpp:
```diff
@@ -6375,12 +6375,21 @@ bool CacheIRCompiler::emitCompareBigIntStringResult(JSOp op,
+#ifndef _MSC_VER
       constexpr auto NotEqual = EqualityKind::NotEqual;
+#else
+      constexpr auto NotEqual = static_cast<bool>(EqualityKind::NotEqual);
```
same comment as 6382 should appear here.

jseyster replied:
Done.

jseyster committed:
Responding to comments from Billy

jseyster committed:
Moved setJitEntryIfNull to its own file

jseyster reviewed (commented):
Thanks for the fast review! I pushed two commits addressing your comments.

mdashti commented on js/src/debugger/Object.cpp:
```diff
@@ -21,6 +21,7 @@
 #include "builtin/Array.h"       // for NewDenseCopiedArray
 #include "builtin/Promise.h"     // for PromiseReactionRecordBuilder
 #include "debugger/Debugger.h"   // for Completion, Debugger
+#include "debugger/Environment.h"
```
Can you please add a comment as to why this header is needed? Similar to other imports in this file.

BillyDonahue replied:
I kind of dislike those comments and Google style discourages them because they are almost immediately out of date.

jseyster replied:
Good eye catching that this header should have a comment to match the style of the others. I tend to agree with Billy, though, that the comment would not be so helpful here. The classes that we need from Environment.h are not directly used in this file, anyway.

mdashti commented on js/src/frontend/BytecodeEmitter.cpp:
```diff
@@ -9771,8 +9771,9 @@ bool BytecodeEmitter::emitPrivateMethodInitializers(ClassEmitter& ce,
     }
-    if (!storedMethodName.append(
-            accessorType == AccessorType::Getter ? ".getter" : ".setter")) {
+    if (!((accessorType == AccessorType::Getter)
+              ? storedMethodName.append(".getter")
+              : storedMethodName.append(".setter"))) {
```
Interesting. Does MSVC get confused by this?

jseyster replied:
Yes. I might have forgotten to mention that the git commit comments include details about some of the changes. From the comment for this change:

"Simplify calls to mozjs StringBuffer::append(const char (&array)[ArrayLength])

"In Clang/g++, type deduction figures out that both sides of the ternary operator (see the diff) can be a char[8] array, which allows overload resolution to choose the StringBuffer::append(const char (&array)[ArrayLength]) overload, where ArrayLength is a template parameter with value 8.

"MSVC's type deduction cannot make this inference, however, so we need to interchange the outer StringBuffer::append() call with the inner ternary operator, simplifying inference. The resulting code is arguably more robust: before this change, the code was only legal because the ".getter" and ".setter" strings happen to have the same length."

BillyDonahue commented on js/src/wasm/WasmBaselineCompile.cpp:
```diff
@@ -2264,12 +2264,10 @@ class BaseStackFrame final : public BaseStackFrameAllocator {
-      StoredFloat32(float f32) : f32(f32) {}
-    } bits{imm};
+      int32_t i32;
+    } bits{imm};  // Note: the default constructor uses the first (f32) member.
```
```suggestion
    } bits{imm};
```
The note is incorrect as this isn't a default constructor.
Could fix it but I think maybe we don't need a note for the language feature?

jseyster replied:
I deleted the comment.

BillyDonahue commented on js/src/vm/Interpreter.cpp:
```diff
@@ -1882,31 +1882,55 @@ void js::ReportInNotObjectError(JSContext* cx, HandleValue lref,
+#  define INTERPRETER_LOOP()
+#  define CASE_NOT_COMPUTED(OP) label_##OP:
+#  define CASE(OP) label_##OP:
+#  define DEFAULT() \
```
These indents didn't need to change. I would like to minimize the diff if possible.

jseyster replied:
Done.

BillyDonahue commented on js/src/wasm/WasmCode-platform.cpp:
```diff
@@ -0,0 +1,39 @@
+#  include <winnt.h>  // For InterlockedCompareExchange64
+#endif
+
+using namespace js::wasm;
```
The using namespace directive can be dangerous.
I don't think this file needs it?

jseyster replied:
We need access to the {{js::wasm}} namespace to resolve {{JumpTable}} in the definition of {{JumpTables::setJitEntryIfNull}} below.

Instead of a {{using}} directive, I updated this code to explicitly place the function definition in the namespace, which I think will be safer.

mdashti commented on js/src/gc/RootMarking.cpp:
```diff
@@ -48,7 +48,16 @@ using RootEnum = RootedValueMap::Enum;
 struct ConcreteTraceable {
   ConcreteTraceable() = delete;
-  void trace(JSTracer*) { MOZ_CRASH("This path is unreachable."); }
+#ifdef _MSC_VER
+  // There should be no callers to ConcreteTraceable::trace(). In clang/gcc, it
+  // is possible to ensure that using the "= delete" syntax, but MSVC is more
+  // aggressive about what is considered a caller.
+  void trace(JSTracer*) {
+    MOZ_ASSERT_UNREACHABLE();
+  }
+#else
+  void trace(JSTracer*) = delete;
+#endif
```
```suggestion
void trace(JSTracer*) { MOZ_ASSERT_UNREACHABLE(); }
```

I'm pretty sure that `= delete` also leads to a compilation issue on GCC. Isn't it better to just use a default unreachable implementation?

BillyDonahue replied:
No it isn't better *IF* it can be avoided. A use of an `=delete` function cannot be compiled.
Maybe I don't know exactly what you mean by a "default unreachable implementation". This macro is a runtime failure, so that's not as good as a compile failure. But if there is actually a use of the function in some way that is implementation-dependent, then I agree we shouldn't leave it to chance.

I'm curious about how this ended up being a compiler-dependent issue, but not enough to block the PR.

jseyster replied:
Agreed that this isn't ideal. The advantage of using the #ifdef is that our non-Windows builders continue to statically validate that there is no attempt to call this deleted function, which gives us confidence that we'll never reach this MOZ_ASSERT_UNREACHABLE() on the Windows build.

I haven't encountered any issue with this deleted function in gcc: it compiles correctly in Evergreen for me.

Unfortunately, I wasn't able to narrow this down to a simple enough case that I could definitively pinpoint the source of the problem. MSVC is more conservative about which call sites it think may be calling this function. From my inspection, I feel confident that this function cannot be called and that the compiler error is a false alarm, but I will have to go back and run the compiler again to get the exact message. (That will take me an hour or so to do, because my Windows Evergreen instance somehow became inaccessible :( ).

mdashti reviewed (approved):
Thanks @jseyster. Beside some minor comments (and the comments already added by Billy), looks good to me.

mdashti committed:
Merge branch 'spidermonkey-esr91.3-cpp-only-merged-prs' into jseyster/windows-cpp20

BillyDonahue commented:
As far as clant-format and indentation. Yes. I think especially for vendored products like this, we should be VERY aggressively limiting diff size so that we can upstream our patches if needed, and to facilitate cleaner future merges as upstream evolves.

I don't personally like indenting after the `#`. It can make preprocessor directives interleave with non-directive lines of code, and it means that adding an outer #if/#endif will cause interior directives to shift, which causes noise later.

jseyster committed:
Second round of comments

jseyster reviewed (commented):
I pushed a new commit responding to the remaining comments.

Note that Mohammad pushed a merge commit to this branch, so some unrelated changes will appear in the "Files changed" section. I recommend looking specifically at the new commit to see the recent changes to the branch. Thanks!

jseyster committed:
Adapt endianness check for MSVC

jseyster commented on mfbt/EndianUtils.h:
```diff
@@ -96,6 +101,14 @@
 #    error "Can't handle mixed-endian architectures"
 #  endif
+#elif defined(_MSC_VER)
+// As of this writing, MSVC only targets little-endian architectures.
```
I don't have a definitive source for this, but it should at least apply to platforms that MongoDB supports.

Additionally, the "possible implementation" of {{std::endian}} strongly implies that Windows targets expect a little-endian architecture.
https://en.cppreference.com/w/cpp/types/endian

Once we are fully transitioned to C++20, we'll have the static_assert to keep this statement honest.

mdashti replied:
@jseyster This change looks good to me.

jseyster reviewed (commented):
I modified EndianUtils.h to detect MSVC in order to address this comment:
https://github.com/10gen/mongo/pull/2878#discussion_r807333109

The proposed implementation will work in C++17 with additional checking enabled in C++20.

jseyster committed:
Do not include version header in C++17

jseyster committed:
Fix build errors

jseyster commented on js/src/wasm/moz.build:
```diff
@@ -48,6 +48,9 @@ UNIFIED_SOURCES += [
     "WasmValue.cpp",
 ]
+SOURCES += [
+    "WasmCode-platform.cpp",
```
In the previous commit where I spun `WasmCode-platform.cpp` into a new file, I forgot to add it to the build system, which broke compilation outside of the `mongo` code base.

I'm adding it here instead of directly to `UNIFIED_SOURCES`, because adding it to `UNIFIED_SOURCES` would require us to update the `src/third_party/mozjs/platform/<ARCH>/<PLATFORM>/build/Unified_cpp_js_src_wasm*.cpp` files for every platform.

@mdashti Unless we decide to bite the bullet and add this to the `UNIFIED_SOURCES` list, we'll have to also add this file to `src/third_party/mozjs/SConstruct`.

mdashti replied:
@jseyster IMO, it's OK to keep it outside `UNIFIED_SOURCES`, as it requires a lot of effort to regenerate the configurations for all platform.

jseyster reviewed (commented):
Posting two more patches to address issues I hit while attempting a standalone build of Spidermonkey.

mdashti committed:
Fixed some minor namespace issues

jseyster commented:
Thanks for all the helpful feedback on this change! I rebased this branch, cleaned up the commits, and pushed everything to the main branch: `spidermonkey-esr91.3-cpp-only`. I'm closing this out as completed.
